cli/src/commands/chmod.rs

```diff
@@ -60,30 +60,37 @@ pub(crate) fn cmd_chmod(
     };
 
     let mut workspace_command = command.workspace_helper(ui)?;
-    // TODO: migrate to .parse_file_patterns()?.to_matcher()
-    let repo_paths: Vec<_> = args
-        .paths
-        .iter()
-        .map(|path| workspace_command.parse_file_path(path))
-        .try_collect()?;
     let commit = workspace_command.resolve_single_rev(&args.revision)?;
     workspace_command.check_rewritable([commit.id()])?;
+    let tree = commit.tree()?;
+    // TODO: No need to add special case for empty paths when switching to
+    // parse_union_filesets(). paths = [] should be "none()" if supported.
+    let fileset_expression = workspace_command.parse_file_patterns(&args.paths)?;
+    let matcher = fileset_expression.to_matcher();
+
+    let unmatched_explicit_paths = fileset_expression
+        .explicit_paths()
+        .filter(|path| tree.path_value(path).is_absent())
+        .map(|path| workspace_command.format_file_path(path))
```

Comment:
This is mostly the same as `cmd_diff` save for the filter.  Perhaps this code could be shared?

Reply:
This version could be extracted to `cli_util`, but I'm not yet sure if the shared function should take `trees: &[_]`. If the caller loops over multiple trees, it's probably better to filter out explicit paths there.

Reply:
I think the shared code could take a `filter` fn, which would be `|path| tree1...is_absent() && tree2...is_absent()` for the other case.

Reply:
Hmm, I think `tree.path_value(path).is_absent()` is the code which we want to remove from the caller. Other iteration and path formatting logic is easy.

If many trees are involved, and if the performance matters, unmatched paths will be updated incrementally.
```rust
let mut unmatched_paths = ..;
for tree in .. {
    for (path, value) in tree.entries_matching(matcher) {
        if !unmatched_paths.is_empty() {
            for p in path.ancestors() {
                unmatched_paths.remove(p);
            }
        }
        ... main logic ...
    }
}

... warn if unmatched_paths isn't empty ...
```

Reply:
Extracted as a function that takes `trees` iterator. As I said, the filter condition is the part which I want to encapsulate in the shared function.